### DISCUSSION TITLE
What is the difference between `/var/lib/containers/storage/tmp` and `/var/tmp`?

### DISCUSSION THREAD

**Discussion by BlackHole1** [containers/podman#20576] [Q&A] [answered]
I noticed that there is a tmp directory under `/var/lib/containers/storage/`, but I couldn't find any documentation or information about this directory. The `imageCopyTmpDir` defaults to using the value of the `TMPDIR` environment variable (or sets it to `/var/tmp` if it's not defined). I would like to know if it's possible to set `imageCopyTmpDir` to `/var/lib/containers/storage/tmp` and if there would be any issues if it's set to `/tmp` or `/run/tmp`.

**Comment by BlackHole1** [comment]
PTAL @rhatdan 🙇‍♂️

**Comment by rhatdan** [comment]
I don't believe it is used.

**Reply by rhatdan** [reply]
I believe https://github.com/containers/podman/pull/20635/files should remove the directory if it exists and remove it, if it was precreated.

**Reply by BlackHole1** [reply]
Thank you for your answer!

**Comment by BlackHole1** [comment]
Another question: Can I set the `imageCopyTmpDir` path to `/run` directory? I’m not sure if there will be any instability issue if its contents are not preserved after each system startup. @rhatdan

**Reply by rhatdan** [reply]
I don;t think that would be a good idea, since this is space limited, and some images are huge.  Also not sure if you could do this other then potentially setting the TMPDIR environment to point at a subdir of /run/

**Reply by BlackHole1** [reply]
Sorry for the confusion in my previous description. What I meant to say is: If the files in the ImageCopyTmpDir directory are automatically deleted after each system reboot, will there be any issues (similar to the **mechanism** of the `/run` directory)?


I just saw the discussion in https://github.com/containers/podman/pull/20635, which raises another question. If I set `ImageCopyTmpDir` to `storage`, it seems that the content in `/var/lib/containers/storage/tmp` will **never be deleted**. I’m not sure if this is reasonable (maybe it should be changed to: `d /var/lib/containers/storage/tmp 0777 root root 30d`).

**Comment by rhatdan** [comment]
Doesn't 
D! /var/lib/containers/storage/tmp 0777 root root 30d

Delete all files in /var/lib/containers/storage/tmp on reboot?

**Reply by BlackHole1** [reply]
> Delete all files in /var/lib/containers/storage/tmp on reboot?

Yes. Actually, my initial requirement was for the virtual machine to restore its original state every time it restarts (except for podman images and containers).
So I need to understand the purpose of each directory in podman (although there is limited documentation on this aspect).

In the current implementation, every time I restart the virtual machine, the contents of ImageCopyTmpDir (/var/tmp) are deleted. So far, I haven’t encountered any issues, but I am concerned that there might be some problems that I have overlooked.

**Reply by BlackHole1** [reply]
> Doesn't
D! /var/lib/containers/storage/tmp 0777 root root 30d

Will this cause the folder to grow indefinitely? Is this expected?

**Reply by rhatdan** [reply]
Folder should only grow if something fails during an action, ordinarily temporary files are deleted.  With this command, the files that are older then 30 days will be deleted on reboot.

**Reply by BlackHole1** [reply]
Thank you for your response, which has helped me understand the purpose of `ImageCopyTmpDir`. I will explain my understanding, and please correct me if I am wrong.

If the *TMPDIR* environment variable is not set, `ImageCopyTmpDir` will use the `/var/tmp` directory, which is managed by *tmpfile.d*, and files inside will be deleted after 30 days. However, if `ImageCopyTmpDir` is set as storage, it will use the `/var/lib/containers/storage/tmp` directory, and any file content in this directory will never be automatically deleted (because *tmpfile.d* does not have the Age field set for this folder by default).

`ImageCopyTmpDir` will only write certain files to the directory in the event of certain failures. Based on this, I can safely delete the contents of the `ImageCopyTmpDir` directory (for example, after the system starts up).

**Reply by rhatdan** [reply]
My understanding of the tmpfile.d is without the Age field, all content will be deleted on a Reboot. I believe you can safely delete the contents at system start, or whenever you don't have any podman commands running.

**Reply by BlackHole1** [reply]
Thank you so much for your answer!

**Comment by BlackHole1** [comment, marked as answer]
https://github.com/containers/podman/pull/20635#issuecomment-1803610085
